### PR TITLE
Add C# Json serializer/de-serializer

### PR DESCRIPTION
**Pull request by YangSan0622**

### CODE REVIEW

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/memory/AwsSchemaRegistryMemeoryLeakDetectionTests.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                { GlueSchemaRegistryConstants.DataFormatType, GlueSchemaRegistryConstants.DataFormat.PROTOBUF },
+            };
+
+            KafkaSerializer.Configure(configs);
+            KafkaDeserializer.Configure(configs);
+
+            var serialized = KafkaSerializer.Serialize(message, message.Descriptor.FullName);
+
+            var deserializedObject =
+                KafkaDeserializer.Deserialize(message.Descriptor.FullName, serialized);
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/memory/AwsSchemaRegistryMemeoryLeakDetectionTests.cs:59]: We should call all the methods of serializer and deserializer to ensure 100% coverage. Add `getSchema` and `canDecode` methods.

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/memory/AwsSchemaRegistryMemeoryLeakDetectionTests.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            var configs = new Dictionary<string, dynamic>
+            {
+                { GlueSchemaRegistryConstants.DataFormatType, GlueSchemaRegistryConstants.DataFormat.JSON },
+            };
+
+            KafkaSerializer.Configure(configs);
+            KafkaDeserializer.Configure(configs);
+
+            var serialized = KafkaSerializer.Serialize(message, "test-topic-json");
+            var deserializedObject = KafkaDeserializer.Deserialize("test-topic-json", serialized);
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/memory/AwsSchemaRegistryMemeoryLeakDetectionTests.cs]: Same as above.

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/memory/AwsSchemaRegistryMemeoryLeakDetectionTests.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        
+        [Test]
+        public void KafkaSerDeMemoryTestForJson()
+        {
+            for (; ;)
+            {
+                SerializeDeserializeJsonMessage();
+            }
+        }
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/memory/AwsSchemaRegistryMemeoryLeakDetectionTests.cs]: Won't only one of them would execute? Shouldn't we combine all of them?

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/serializer/GlueSchemaRegistryKafkaSerializerTests.cs [hunk 4]
```diff
@@ -117,17 +137,37 @@ public void KafkaSerDeTestForAllProtobufTypes(IMessage message)
...
+            var configs = new Dictionary<string, dynamic>
+            {
+                { GlueSchemaRegistryConstants.DataFormatType, GlueSchemaRegistryConstants.DataFormat.JSON },
+            };
+
+            KafkaSerializer.Configure(configs);
+            KafkaDeserializer.Configure(configs);
+
+            var serialized = KafkaSerializer.Serialize(message, "test-topic-json");
+            var deserializedObject = KafkaDeserializer.Deserialize("test-topic-json", serialized);
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/serializer/GlueSchemaRegistryKafkaSerializerTests.cs:142]: Also assert the `getSchema` and `canDecode` methods. Let's cover all the methods.

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+        private static byte[] WriteBytes(JsonNode data)
+        {
+            var options = new JsonWriterOptions();
+            using var stream = new MemoryStream();
+            using var writer = new Utf8JsonWriter(stream, options);
+
+            data.WriteTo(writer, new JsonSerializerOptions());
+            writer.Flush();
+            var array = stream.ToArray();
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs:93]: Shouldn't we close the stream?

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        private static JsonSchema GetSchema(object data)
+        {
+            try
+            {
+                if (IsWrapper(data))
+                {
+                    var dataObject = (JsonDataWithSchema)data;
+                    var getSchemaTask = JsonSchema.FromJsonAsync(dataObject.Schema);
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs:105]: Why Async?

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                    return getSchemaTask.Result;
+                }
+                var schema = data is string dataJsonString
+                    ? JsonSchema.FromSampleJson(dataJsonString)
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs:110]: What's the significance of this?

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+                return schemaNode;
+            }
+            catch (Exception e)
+            {
+                throw new AwsSchemaRegistryException(
+                    "Could not generate schema from the type provided " + data.GetType(), e);
+            }
+        }
+
+        private static JsonNode GetDataNode([NotNull] object data)
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs:144]: Please test compatibility between Java and C#,

Java serialized data can be de-serialized by C# and vice-versa.

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/memory/AwsSchemaRegistryMemeoryLeakDetectionTests.cs (hunk 3 above)
Comment (YangSan0622) [reply native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/memory/AwsSchemaRegistryMemeoryLeakDetectionTests.cs]: Good idea, I will combine them

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs (hunk 5 above)
Comment (YangSan0622) [reply native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs:93]: good catch, fixed

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs (hunk 6 above)
Comment (YangSan0622) [reply native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs:105]: Somehow this library does not have a sync version of this method.

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs (hunk 7 above)
Comment (YangSan0622) [reply native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs:110]: I am trying to support one more use case which is when the data object passed in is a plain json string. Do you think I should keep it?

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs (hunk 8 above)
Comment (YangSan0622) [reply native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe/serializer/json/JsonSerializer.cs:144]: I have verified it, the serialized bytes are the same.

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/deserializer/GlueSchemaRegistryKafkaDeserializerTests.cs [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        public void TestDeserializerWithMessageNotEncodedBySerializer_Avro()
+        {
+            var avroRecord = RecordGenerator.GetTestAvroRecord();
+
+            var avroSerializer = new AvroSerializer();
+            var bytes = avroSerializer.Serialize(avroRecord);
+
+            Assert.DoesNotThrow(() => Deserializer.CanDecode(bytes));
+            Assert.False(Deserializer.CanDecode(bytes));
+            Assert.Throws(typeof(AwsSchemaRegistryException), () => Deserializer.DecodeSchema(bytes));
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/deserializer/GlueSchemaRegistryKafkaDeserializerTests.cs]: Minor: Assert message as well.

File: native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/deserializer/GlueSchemaRegistryKafkaDeserializerTests.cs [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+                { GlueSchemaRegistryConstants.DataFormatType, GlueSchemaRegistryConstants.DataFormat.PROTOBUF },
+            };
+
+            KafkaSerializer.Configure(configs);
+
+            var bytes = KafkaSerializer.Serialize(protobufMessage, protobufMessage.Descriptor.FullName);
+
+            Assert.DoesNotThrow(() => Deserializer.CanDecode(bytes));
+            Assert.True(Deserializer.CanDecode(bytes));
+            Assert.DoesNotThrow(() => Deserializer.DecodeSchema(bytes));
```
Comment (blacktooth) [inline native-schema-registry/csharp/AWSGsrSerDe/AWSGsrSerDe.Tests/deserializer/GlueSchemaRegistryKafkaDeserializerTests.cs:148]: How about the `Decode` method?

**Review by blacktooth — APPROVED** [review]
Please push after adding the requested changes.

**Pull request merged**